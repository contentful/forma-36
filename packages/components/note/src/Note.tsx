import { cx } from 'emotion';
import React from 'react';
import { Flex, Grid } from '@contentful/f36-core';
import type {
  CommonProps,
  PropsWithHTMLElement,
  ExpandProps,
} from '@contentful/f36-core';
import { Button } from '@contentful/f36-button';
import { Heading, Paragraph, Text } from '@contentful/f36-typography';
import {
  CheckCircleIcon,
  CloseIcon,
  ErrorCircleIcon,
  InfoCircleIcon,
  WarningIcon,
} from '@contentful/f36-icons';
import { Icon } from '@contentful/f36-icon';

import { getNoteStyles } from './Note.styles';

const icons = {
  primary: InfoCircleIcon,
  positive: CheckCircleIcon,
  negative: ErrorCircleIcon,
  warning: WarningIcon,
  neutral: InfoCircleIcon,
};

export type NoteVariant =
  | 'negative'
  | 'positive'
  | 'primary'
  | 'warning'
  | 'neutral';

export type NoteInternalProps = CommonProps & {
  /**
   * Determines style variation of Note component
   */
  variant?: NoteVariant;
  /**
   * Sets title in the Note
   */
  title?: React.ReactNode;
  /**
   * Children of Note
   */
  children?: React.ReactNode | string;
  /**
   * Defines if the close button should be rendered
   * @default false
   */
  withCloseButton?: boolean;
  /**
   * Callback for handling closing
   */
  onClose?: React.MouseEventHandler<HTMLButtonElement>;
};
export type NoteProps = PropsWithHTMLElement<NoteInternalProps, 'article'>;

/**
 * @description: Note provides context and information about a situation or action.
 */
export const Note = React.forwardRef<HTMLElement, ExpandProps<NoteProps>>(
  (props, ref) => {
    const {
      children,
      className,
      withCloseButton = false,
      variant = 'primary',
      onClose,
      testId = 'cf-ui-note',
      title,
      ...otherProps
    } = props;

    const styles = getNoteStyles();

    return (
      <Grid
        {...otherProps}
        columns={withCloseButton ? 'auto 1fr 24px' : 'auto 1fr'} // 24px is the width of the close button
        as="article"
        className={cx(styles.container({ variant }), className)}
        testId={testId}
        ref={ref}
        padding="spacingM"
      >
        <Icon
          as={icons[variant]}
          variant={variant === 'neutral' ? 'muted' : variant}
          size={title ? 'medium' : 'small'}
        />
        <Flex flexDirection="column">
          {title && (
            <Heading
              as="h2"
              className={styles.title}
              marginBottom={!children ? 'none' : 'spacingS'}
            >
              {title}
            </Heading>
          )}
          {children && (
            <Text
              as="div"
              lineHeight="lineHeightM"
              className={styles.description}
            >
              {typeof children === 'string' ? (
                <Paragraph marginBottom="none">{children}</Paragraph>
              ) : (
                children
              )}
            </Text>
          )}
        </Flex>
        {withCloseButton && (
          <Button
            variant="transparent"
            startIcon={<CloseIcon className={styles.closeIcon} />}
            onClick={onClose}
            testId={`${testId}-close`}
            aria-label="Dismiss"
            className={styles.close}
          />
        )}
      </Grid>
    );
  },
);

Note.displayName = 'Note';
