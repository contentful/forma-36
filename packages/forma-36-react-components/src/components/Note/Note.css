@import 'settings/colors';
@import 'settings/typography';
@import 'settings/dimensions';
@import 'settings/borders';

.Note {
  padding: var(--spacing-m);
  margin: 0;
  font-family: var(--font-stack-primary);
  font-weight: var(--font-weight-normal);
  line-height: var(--line-height-default);
  font-size: var(--font-size-m);
  color: var(--color-text-base);
  display: flex;
  flex-direction: row;
  border-radius: var(--border-radius-medium);
}

.Note__title {
  font-weight: var(--font-weight-medium);
  font-size: var(--font-size-l);
  margin-bottom: var(--spacing-xs);
  color: var(--color-text-dark);
  line-height: var(--line-height-l);
}

.Note__content {
  width: 100%;
  line-height: var(--line-height-m);
}

.Note--primary {
  background-color: var(--color-blue-lightest);
  border: 1px solid var(--color-blue-light);
}

.Note--primary a {
  color: var(--color-blue-dark);
}

.Note--negative {
  background-color: var(--color-red-lightest);
  border: 1px solid var(--color-red-light);
}

.Note--negative a {
  color: var(--color-red-dark);
}

.Note--positive {
  background-color: var(--color-green-lightest);
  border: 1px solid var(--color-green-light);
}

.Note--positive a {
  color: var(--color-green-dark);
}

.Note--warning {
  background: var(--color-orange-lightest);
  border: 1px solid var(--color-orange-light);
}

.Note--warning a {
  color: var(--color-orange-dark);
}

.Note--hasCloseButton {
  padding-right: var(--spacing-s);
}

.Note--hasCloseButton .Note__info {
  margin-right: var(--spacing-s);
}

.Note__icon {
  margin-right: var(--spacing-s);
  display: flex;
}

.Note__dismiss {
  background: transparent;
  border: none;
  cursor: pointer;
  outline: none;
  margin: 0 0 auto auto;
  pointer-events: all;
  display: flex;
}
