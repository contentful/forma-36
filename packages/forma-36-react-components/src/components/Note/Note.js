import React from 'react';
import classNames from 'classnames';
import PropTypes from 'prop-types';
import Icon from '../Icon';
import { iconName } from '../Icon/constants';
import styles from './Note.css';

export const NoteType = {
  PRIMARY: 'primary',
  POSITIVE: 'positive',
  NEGATIVE: 'negative',
  WARNING: 'warning',
};

const Icons = {
  [NoteType.PRIMARY]: iconName.InfoCircle,
  [NoteType.POSITIVE]: iconName.CheckCircle,
  [NoteType.NEGATIVE]: iconName.Warning,
  [NoteType.WARNING]: iconName.Warning,
};

export default function Note(props) {
  const icon = Icons[props.noteType];
  if (!icon) {
    throw new Error(
      `Intent ${props.noteType} is not supported in Note component.`,
    );
  }

  return (
    <div
      style={props.style}
      className={classNames(styles.Note, props.extraClassNames, {
        [styles['Note--primary']]: props.noteType === NoteType.PRIMARY,
        [styles['Note--positive']]: props.noteType === NoteType.POSITIVE,
        [styles['Note--negative']]: props.noteType === NoteType.NEGATIVE,
        [styles['Note--warning']]: props.noteType === NoteType.WARNING,
      })}
      data-test-id={props.testId}
    >
      <div className={styles.Note__icon}>
        <Icon icon={icon} color={props.noteType} />
      </div>
      <div className={styles.Note__info}>
        {props.title && <div className={styles.Note__title}>{props.title}</div>}
        <div>{props.children}</div>
      </div>
    </div>
  );
}

Note.Type = NoteType;

Note.propTypes = {
  title: PropTypes.string,
  noteType: PropTypes.oneOf([
    NoteType.POSITIVE,
    NoteType.NEGATIVE,
    NoteType.PRIMARY,
    NoteType.WARNING,
  ]),
  children: PropTypes.node.isRequired,
  testId: PropTypes.string,
  extraClassNames: PropTypes.string,
  style: PropTypes.object,
};

Note.defaultProps = {
  title: undefined,
  noteType: NoteType.PRIMARY,
  testId: 'cf-ui-note',
  extraClassNames: undefined,
  style: undefined,
};
