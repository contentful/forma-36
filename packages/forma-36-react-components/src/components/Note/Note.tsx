import React from 'react';
import classNames from 'classnames';
import type { CSSProperties } from 'react';

import Icon, { IconType } from '../Icon';
import { iconName } from '../Icon/constants';
import styles from './Note.css';
import IconButton from '../IconButton';

const Icons = {
  primary: iconName.InfoCircle,
  positive: iconName.CheckCircle,
  negative: iconName.ErrorCircle,
  warning: iconName.Warning,
};

export interface NoteProps {
  noteType?: 'primary' | 'positive' | 'negative' | 'warning';
  className?: string;
  title?: string;
  style?: CSSProperties;
  testId?: string;
  children: React.ReactNode;
  hasCloseButton?: boolean;
  onClose?: Function;
}

export function Note({
  children,
  className,
  hasCloseButton,
  noteType = 'primary',
  onClose,
  style,
  testId = 'cf-ui-note',
  title,
}: NoteProps): React.ReactElement {
  const icon = Icons[noteType!] as IconType; // eslint-disable-line @typescript-eslint/no-non-null-assertion
  if (!icon) {
    throw new Error(`Intent ${noteType} is not supported in Note component.`);
  }

  return (
    <div
      style={style}
      className={classNames(styles.Note, className, {
        [styles[`Note--${noteType}`]]: noteType,
        [styles['Note--hasCloseButton']]: hasCloseButton,
      })}
      data-test-id={testId}
    >
      <div className={styles.Note__icon}>
        <Icon icon={icon} color={noteType} size={title ? 'medium' : 'small'} />
      </div>
      <div className={styles.Note__info}>
        {title && <div className={styles.Note__title}>{title}</div>}
        <div className={styles.Note__content}>{children}</div>
      </div>
      {hasCloseButton && (
        <IconButton
          buttonType="secondary"
          iconProps={{ icon: 'Close' }}
          onClick={() => {
            if (onClose) {
              onClose();
            }
          }}
          testId="cf-ui-note-close"
          label="Dismiss"
          className={styles.Note__dismiss}
        />
      )}
    </div>
  );
}

export default Note;
