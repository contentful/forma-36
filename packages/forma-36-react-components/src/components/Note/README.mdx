---
title: 'Note'
type: 'component'
status: 'stable'
slug: /components/note/
github: 'https://github.com/contentful/forma-36/tree/master/packages/forma-36-react-components/src/components/Note'
storybook: 'https://f36-storybook.contentful.com/?path=/story/components-note--default'
---

Notes provide context and information about a situation or action. Notes draw the users attention to a circumstance relevant to an action they're performing or a screen they're accessing. Don't confuse notes with notifications, which appear as feedback after a user has performed an action.

## Table of contents

- [How to use Note](#how-to-use-note)
- [Component variations](#component-variations)
- [Code examples](#code-examples)
- [Content recommendations](#content-recommendations)
- [Accessibility](#accessibility)

## How to use Note

- Use an adequate note type to communicate the right kind of information
- Place the note visually closer to the action or real estate it relates to
- Unlike notifications, notes don't show up as feedback only after a user has performed an action

## Component variations

Give couple of clear examples of the usage of the components, that includes import

## Code examples

The Note component can be configured in a number of different ways. Learn when to use certain variations with this guide.

### Basic example

```jsx
<React.Fragment>
  <Note>
    Information that is relevant to the context the user is currently in.
  </Note>
</React.Fragment>
```

### Note type (`noteType`)

```jsx
<React.Fragment>
  <Note>Primary (default)</Note>
  <Note noteType="positive">Positive</Note>
  <Note noteType="warning">Warning</Note>
  <Note noteType="negative">Negative</Note>
</React.Fragment>
```

### Note with title (`title`)

```jsx
<React.Fragment>
  <Note title="Primary title">Primary (default)</Note>
  <Note title="Positive title" noteType="positive">
    Positive
  </Note>
  <Note title="Warning title" noteType="warning">
    Warning
  </Note>
  <Note title="Negative title" noteType="negative">
    Negative
  </Note>
</React.Fragment>
```

## Content recommendations

- Use clear and succinct copy
- Ensure the copy is easily scannable. Focus on the objective and limit the number of concepts in each sentence
- Use progressive disclosure. Eliminate details that can be revealed as the user interacts and requires more information
- Use active voice, present tense, and consider tone of voice depending on the circumstance
- To add additional context, link out to documentation
- Do not preface the instructions with introductory text, such as "please"

## Accessibility

Missing
