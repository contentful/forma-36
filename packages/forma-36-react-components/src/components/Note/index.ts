export * from './Note';
export { default } from './Note';
