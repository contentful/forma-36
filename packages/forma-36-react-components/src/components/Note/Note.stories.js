import React from 'react';
import { storiesOf } from '@storybook/react';
import { host } from 'storybook-host';
import { withNotes } from '@storybook/addon-notes';
import { text, select } from '@storybook/addon-knobs';
import NoteNotes from './Note.md';
import Note from './Note';

storiesOf('Components|Note', module)
  .addDecorator(
    host({
      align: 'center top',
      cropMarks: false,
    }),
  )
  .add(
    'default',
    withNotes({ markdown: NoteNotes })(() => (
      <Note
        noteType={select(
          'noteType',
          [
            Note.Type.POSITIVE,
            Note.Type.NEGATIVE,
            Note.Type.WARNING,
            Note.Type.PRIMARY,
          ],
          Note.defaultProps.noteType,
        )}
        title={text('title', '')}
      >
        {text(
          'children',
          'A piece of information that is relevant to the context the user is currently in.',
        )}
      </Note>
    )),
  )
  .add(
    'all notes',
    withNotes({ markdown: NoteNotes })(() => (
      <div>
        <Note>
          A piece of information that is relevant to the context the user is
          currently in.
        </Note>
        <div style={{ marginTop: 20 }} />
        <Note noteType={Note.Type.NEGATIVE}>
          A piece of information that is relevant to the context the user is
          currently in.
        </Note>
        <div style={{ marginTop: 20 }} />
        <Note noteType={Note.Type.WARNING}>
          A piece of information that is relevant to the context the user is
          currently in.
        </Note>
        <div style={{ marginTop: 20 }} />
        <Note noteType={Note.Type.POSITIVE}>
          A piece of information that is relevant to the context the user is
          currently in.
        </Note>
        <div style={{ marginTop: 60 }} />
        <Note title="Short, yet succint title">
          A piece of information that is relevant to the context the user is
          currently in. If you like it then you should put{' '}
          <a href="https://contentful.com">a link</a> in it.
        </Note>
        <div style={{ marginTop: 20 }} />
        <Note noteType={Note.Type.NEGATIVE} title="Short, yet succint title">
          A piece of information that is relevant to the context the user is
          currently in. If you like it then you should put{' '}
          <a href="https://contentful.com">a link</a> in it.
        </Note>
        <div style={{ marginTop: 20 }} />
        <Note noteType={Note.Type.WARNING} title="Short, yet succint title">
          A piece of information that is relevant to the context the user is
          currently in. If you like it then you should put{' '}
          <a href="https://contentful.com">a link</a> in it.
        </Note>
        <div style={{ marginTop: 20 }} />
        <Note noteType={Note.Type.POSITIVE} title="Short, yet succint title">
          A piece of information that is relevant to the context the user is
          currently in. If you like it then you should put{' '}
          <a href="https://contentful.com">a link</a> in it.
        </Note>
      </div>
    )),
  );
