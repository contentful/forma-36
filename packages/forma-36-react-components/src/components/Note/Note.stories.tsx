import React from 'react';
import { storiesOf } from '@storybook/react';
import { boolean, select, text } from '@storybook/addon-knobs';

import notes from './Note.md';
import Note from './Note';

storiesOf('Components|Note', module)
  .addParameters({
    propTypes: Note['__docgenInfo'],
  })
  .add(
    'default',
    () => (
      <Note
        noteType={select(
          'noteType',
          ['positive', 'negative', 'warning', 'primary'],
          Note.defaultProps.noteType,
        )}
        title={text('title', '')}
        hasCloseButton={boolean('hasCloseButton', false)}
      >
        {text(
          'children',
          'A piece of information that is relevant to the context the user is currently in.',
        )}
      </Note>
    ),
    { notes },
  )
  .add(
    'all notes',
    () => (
      <div>
        <Note>
          A piece of information that is relevant to the context the user is
          currently in.
        </Note>
        <div style={{ marginTop: 20 }} />
        <Note noteType={'negative'}>
          A piece of information that is relevant to the context the user is
          currently in.
        </Note>
        <div style={{ marginTop: 20 }} />
        <Note noteType={'warning'}>
          A piece of information that is relevant to the context the user is
          currently in.
        </Note>
        <div style={{ marginTop: 20 }} />
        <Note noteType={'positive'}>
          A piece of information that is relevant to the context the user is
          currently in.
        </Note>
        <div style={{ marginTop: 60 }} />
        <Note title="Short, yet succint title">
          A piece of information that is relevant to the context the user is
          currently in. If you like it then you should put{' '}
          <a href="https://contentful.com">a link</a> in it.
        </Note>
        <div style={{ marginTop: 20 }} />
        <Note noteType={'negative'} title="Short, yet succint title">
          A piece of information that is relevant to the context the user is
          currently in. If you like it then you should put{' '}
          <a href="https://contentful.com">a link</a> in it.
        </Note>
        <div style={{ marginTop: 20 }} />
        <Note noteType={'warning'} title="Short, yet succint title">
          A piece of information that is relevant to the context the user is
          currently in. If you like it then you should put{' '}
          <a href="https://contentful.com">a link</a> in it.
        </Note>
        <div style={{ marginTop: 20 }} />
        <Note noteType={'positive'} title="Short, yet succint title">
          A piece of information that is relevant to the context the user is
          currently in. If you like it then you should put{' '}
          <a href="https://contentful.com">a link</a> in it.
        </Note>
      </div>
    ),
    { notes },
  );
