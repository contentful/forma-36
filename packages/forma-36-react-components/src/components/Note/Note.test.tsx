import React from 'react';
import { shallow, mount } from 'enzyme';
import { axe } from 'jest-axe';
import Note from './Note';

it('renders the component', () => {
  const output = shallow(
    <Note>
      A piece of information that is relevant to the context the user is
      currently in.
    </Note>,
  );

  expect(output).toMatchSnapshot();
});

it('renders the component with title', () => {
  const output = shallow(
    <Note noteType={Note.Type.POSITIVE} title="Positive title">
      A piece of information that is relevant to the context the user is
      currently in.
    </Note>,
  );

  expect(output).toMatchSnapshot();
});

it('renders the component with an additional class name', () => {
  const output = shallow(
    <Note className="my-extra-class">
      A piece of information that is relevant to the context the user is
      currently in.
    </Note>,
  );

  expect(output).toMatchSnapshot();
});

it(`has no a11y issues`, async () => {
  const output = mount(
    <Note>
      A piece of information that is relevant to the context the user is
      currently in.
    </Note>,
  ).html();

  const results = await axe(output);

  expect(results).toHaveNoViolations();
});
