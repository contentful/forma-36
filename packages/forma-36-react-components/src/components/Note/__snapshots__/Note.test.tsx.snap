// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the component 1`] = `
<div
  class="Note Note--primary"
  data-test-id="cf-ui-note"
>
  <div
    class="Note__icon"
  >
    <svg
      class="Icon Icon--small Icon--primary"
      data-test-id="cf-ui-icon"
      height="1em"
      viewBox="0 0 24 24"
      width="1em"
    >
      <path
        d="M0 0h24v24H0z"
        fill="none"
      />
      <path
        d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-6h2v6zm0-8h-2V7h2v2z"
      />
    </svg>
  </div>
  <div
    class="Note__info"
  >
    <div
      class="Note__content"
    >
      A piece of information that is relevant to the context the user is currently in.
    </div>
  </div>
</div>
`;

exports[`renders the component with an additional class name 1`] = `
<div
  class="Note my-extra-class Note--primary"
  data-test-id="cf-ui-note"
>
  <div
    class="Note__icon"
  >
    <svg
      class="Icon Icon--small Icon--primary"
      data-test-id="cf-ui-icon"
      height="1em"
      viewBox="0 0 24 24"
      width="1em"
    >
      <path
        d="M0 0h24v24H0z"
        fill="none"
      />
      <path
        d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-6h2v6zm0-8h-2V7h2v2z"
      />
    </svg>
  </div>
  <div
    class="Note__info"
  >
    <div
      class="Note__content"
    >
      A piece of information that is relevant to the context the user is currently in.
    </div>
  </div>
</div>
`;

exports[`renders the component with title 1`] = `
<div
  class="Note Note--positive"
  data-test-id="cf-ui-note"
>
  <div
    class="Note__icon"
  >
    <svg
      class="Icon Icon--medium Icon--positive"
      data-test-id="cf-ui-icon"
      height="1em"
      viewBox="0 0 24 24"
      width="1em"
    >
      <path
        d="M0 0h24v24H0z"
        fill="none"
      />
      <path
        d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm-2 15l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
      />
    </svg>
  </div>
  <div
    class="Note__info"
  >
    <div
      class="Note__title"
    >
      Positive title
    </div>
    <div
      class="Note__content"
    >
      A piece of information that is relevant to the context the user is currently in.
    </div>
  </div>
</div>
`;
